### PR TITLE
Use saturated part colour when calculating edge colour

### PR DESCRIPTION
**Pull request by trevorsandy**
This update applies the changes captured in #618.

Cheers,

### CODE REVIEW

**Comment by rsbx** [comment]
I've been offline most of day and only looked at the patch briefly but I 
noticed that you include the supersaturation/hue complement calculations 
that I had comment out. I don't think they should be included or at the 
very least they should be discussed first. I had left them in commented 
out mostly as documentation of what is possible; not what is worth while.

I will take a close look at the patch tomorrow.

**Comment by trevorsandy** [comment]
Indeed, I enabled the saturation adjustment for evaluation. It can be trivially excised. 

There is also some dead code in your patch - for example the -= operator and rgb vector assigned at 'Get base color in hSL' are never used.

Cheers,

**Comment by rsbx** [comment]
Assuming you meant the super-saturation and hue adjustment: that's fine; 
it was interesting to try but the results were disappointing. If you 
meant the saturation adjustment in general: it's pretty much the point 
of the patch.

Not surprised about the dead code. I had done all of the initial work in 
Perl with a lot of debugging code that I did not move into C++ for 
LeoCAD. But I didn't want to change it too much in case I had to debug 
it again. Part of the reason I label it 'proof-of-concept' code. Once 
the functionality is agreed on, the code cleanup is pretty easy.

**Comment by trevorsandy** [comment]
Yes, I meant the super-saturation and hue adjustment. I've since commented this block and removed the dead code.

Cheers,

**Comment by rsbx** [comment]
@trevorsandy 

I'm testing and  looking at the patched code and there's some things I'm not sure about:
- It appears the you're using the same control variable, Preferences.mPartColorValueLDIndex, for the Saturation adjustment value and for determining light/dark status. I did that as a hack because I needed a Saturation control and no longer needed the LDindex (and didn't want to learn QT UI programming) but I didn't intend for that to hit master. Continuing to do so also means that when the user adjusts either Saturation or LDindex, the other will be changed in a way that the user did not intend. A separate control variable is warranted.
- How do the Color.Adjusted booleans get cleared? I'm able to adjust Saturation/Contrast multiple times but I don't see how.
- My second patch saved the Edge color read from the color config file to use for Edge luma in light/dark tests but your code uses the current edge color. This becomes a problem when the previously adjusted edge color is exactly the base color. I was not able to demonstrate the problem in practice so either I'm missing something or there's a lurking edge condition but I just wasn't able to trigger __yet__.
- I think the setting of LightDarkControl is confusing.

I also see that you fixed the startup problem when High Contrast and Auto Edge were both enabled; it was on the TODO list to track down.

**Comment by trevorsandy** [comment]
> It appears the you're using the same control variable, Preferences.mPartColorValueLDIndex, for the Saturation adjustment value and for determining light/dark status. I did that as a hack because I needed a Saturation control and no longer needed the LDindex (and didn't want to learn QT UI programming) but I didn't intend for that to hit master. Continuing to do so also means that when the user adjusts either Saturation or LDindex, the other will be changed in a way that the user did not intend. A separate control variable is warranted.

Saturation and LDValue are settings for mutually exclusive behaviour (i.e. High Contrast Stud Style & Automate Edge Colours). The user must simply set the desired value for the behaviour enabled. This is why there is a message if Automate Edge Colour is set, High Contrast Colour settings will be ignored. This behaviour is not going to be so frequently used to require an additional setting just to persist the other while it is not in use.

> How do the Color.Adjusted booleans get cleared? I'm able to adjust Saturation/Contrast multiple times but I don't see how.

I don't understand your question. I have not implemented any booleans to adjust colour or saturation. These settings are controlled by their respective sliders in Preferences. Booleans only enable or disable the aforementioned behaviours.

> My second patch saved the Edge color read from the color config file to use for Edge luma in light/dark tests but your code uses the current edge color. This becomes a problem when the previously adjusted edge color is exactly the base color. I was not able to demonstrate the problem in practice so either I'm missing something or there's a lurking edge condition but I just wasn't able to trigger yet.

Paste some sample code and line numbers to better pinpoint what you are talking about here. Perhaps I followed the code from your initial patch. If it's wrong, highlight what needs to change and I'll change it.

> I think the setting of LightDarkControl is confusing.

How ? Nothing has changed in the user interface between the current implementation of the high contrast stud style settings and your saturation patch.

Cheers,

**Comment by rsbx** [comment]
> > It appears the you're using the same control variable, Preferences.mPartColorValueLDIndex, for the Saturation adjustment value and for determining light/dark status. I did that as a hack because I needed a Saturation control and no longer needed the LDindex (and didn't want to learn QT UI programming) but I didn't intend for that to hit master. Continuing to do so also means that when the user adjusts either Saturation or LDindex, the other will be changed in a way that the user did not intend. A separate control variable is warranted.
> 
> Saturation and LDValue are settings for mutually exclusive behaviour (i.e. High Contrast Stud Style & Automate Edge Colours). The user must simply set the desired value for the behaviour enabled. This is why there is a message if Automate Edge Colour is set, High Contrast Colour settings will be ignored. This behaviour is not going to be so frequently used to require an additional setting just to persist the other while it is not in use.

We disagree here. I foresee myself using Auto Edge during design and for some rendering and High Contrast for some rendering and instructions. The use of different setups will likely be interleaved.

> > How do the Color.Adjusted booleans get cleared? I'm able to adjust Saturation/Contrast multiple times but I don't see how.
> 
> I don't understand your question. I have not implemented any booleans to adjust colour or saturation. These settings are controlled by their respective sliders in Preferences. Booleans only enable or disable the aforementioned behaviours.

In the `lcColor` struct (lc_colors.h) there's a new `bool Ajusted`; in `lcAdjustStudStyleColors` (lc_colors.cpp) it's tested and setting the edge color is skipped in the Auto Edge mode. It appears to be part of your initial #621 patch.

> > My second patch saved the Edge color read from the color config file to use for Edge luma in light/dark tests but your code uses the current edge color. This becomes a problem when the previously adjusted edge color is exactly the base color. I was not able to demonstrate the problem in practice so either I'm missing something or there's a lurking edge condition but I just wasn't able to trigger yet.
> 
> Paste some sample code and line numbers to better pinpoint what you are talking about here. Perhaps I followed the code from your initial patch. If it's wrong, highlight what needs to change and I'll change it.

My initial patch did not save the edge color read from the color file; that's what I fixed with the second patch. In the second patch, look for instances of `Config.Edge` in lc_colors.{cpp,h}.

> > I think the setting of LightDarkControl is confusing.
> 
> How ? Nothing has changed in the user interface between the current implementation of the high contrast stud style settings and your saturation patch.

This is code in question:
`        const float LightDarkControl = Preferences.mAutomateEdgeColor ?
                                Preferences.mPartColorValueLDIndex :
                                LC_SRGB_TO_LINEAR(Preferences.mPartColorValueLDIndex);
`
It's correct. I just think it's confusing (why  LC_SRGB_TO_LINEAR for one path but not the other) and would have done it in a clearer but less efficient manner or just used a separate control variable. Related to reusing the same control variable.

> Cheers,

**Comment by trevorsandy** [comment]
> We disagree here. I foresee myself using Auto Edge during design and for some rendering and High Contrast for some rendering and instructions. The use of different setups will likely be interleaved.

Indeed, and you will have to manipulate the settings (e.g. enable/disable) as you go from HCS to AEC and vice versa. My thinking is, at the time the user is manually 'switching', s/he can simply set the saturation/light-dark value if needed.

> In the lcColor struct (lc_colors.h) there's a new bool Ajusted; in lcAdjustStudStyleColors (lc_colors.cpp) it's tested and setting the edge color is skipped in the Auto Edge mode. It appears to be part of your initial #621 patch.

Ah, yes, Adjusted is cleared whenever the color file is loaded:

```
static std::vector<lcColor> lcParseColorFile(lcFile& File)
{
	char Line[1024], Token[1024];
	std::vector<lcColor> Colors;
	lcColor Color;
        ...
```
The last line of code above will call the lcColor struct constructor which sets the Adjusted (and Transparent) flags to false.

> My initial patch did not save the edge color read from the color file; that's what I fixed with the second patch. In the second patch, look for instances of Config.Edge in lc_colors.{cpp,h}.

Ok, I did not use your patch update code. I opted for the 'Adjusted' flag instead. The color file loads whenever LoadColors() is called, you can see when this is by looking up this call in the source. Once a colour is adjusted, there is no need to adjust it again so the flag allows the lcAdjustStudStyleColors() loop to skip an adjusted colour. However, neither this nor your perceived bug should ever trigger in the current LeoCAD behaviour because lcAdjustStudStyleColors() is only called from lcLoadColorFile() so there should never be an opportunity to adjust an already adjusted colour. I added this flag in the event Leo decides to recognize the !COLOUR LDraw meta command when present in model files in which case I imagine the colour file will not be reloaded. For LPub3D, I use the !COLOUR meta command in my modified version of LeoCAD.

> const float LightDarkControl = Preferences.mAutomateEdgeColor ? Preferences.mPartColorValueLDIndex : LC_SRGB_TO_LINEAR(Preferences.mPartColorValueLDIndex);
It's correct. I just think it's confusing (why LC_SRGB_TO_LINEAR for one path but not the other) 

Ok. As we already discussed, this flag serves both HCS and AEC behaviours so based on what is enabled, the appropriate assignment is handled. AEC uses the index directly while HCS uses linear derivative.

Cheers,

**Comment by rsbx** [comment]
>     We disagree here. I foresee myself using Auto Edge during design and
>     for some rendering and High Contrast for some rendering and
>     instructions. The use of different setups will likely be interleaved.
> 
> Indeed, and you will have to manipulate the settings (e.g. 
> enable/disable) as you go from HCS to AEC and vice versa. My thinking 
> is, at the time the user is manually 'switching', s/he can simply set 
> the saturation/light-dark value if needed.

Like I said, we disagree here.

>     In the lcColor struct (lc_colors.h) there's a new bool Ajusted; in
>     lcAdjustStudStyleColors (lc_colors.cpp) it's tested and setting the
>     edge color is skipped in the Auto Edge mode. It appears to be part
>     of your initial #621 <https://github.com/leozide/leocad/pull/621> patch.
> 
> Ah, yes, Adjusted is cleared whenever the color file is loaded:
> 
> |static std::vector<lcColor> lcParseColorFile(lcFile& File) { char 
> Line[1024], Token[1024]; std::vector<lcColor> Colors; lcColor Color; ... |
> 
> The last line of code above will call the lcColor struct constructor 
> which sets the Adjusted (and Transparent) flags to false.

So, is the color file load whenever these preferences are changed? If 
so, that's new information and I'm OK with your approach; I had assumed 
otherwise.

>     My initial patch did not save the edge color read from the color
>     file; that's what I fixed with the second patch. In the second
>     patch, look for instances of Config.Edge in lc_colors.{cpp,h}.
> 
> Ok, I did not use your patch update code. I opted for the 'Adjusted' 
> flag instead. The color file loads whenever LoadColors() is called, you 
> can see when this is by looking up this call in the source. Once a 
> colour is adjusted, there is no need to adjust it again so the flag 
> allows the lcAdjustStudStyleColors() loop to skip an adjusted colour. 
> However, neither this nor your perceived bug should ever trigger in the 
> current LeoCAD behaviour because lcAdjustStudStyleColors() is only 
> called from lcLoadColorFile() so there should never be an opportunity to 
> adjust an already adjusted colour. I added this flag in the event Leo 
> decides to recognize the !COLOUR LDraw meta command when present in 
> model files in which case I imagine the colour file will not be 
> reloaded. For LPub3D, I use the !COLOUR meta command in my modified 
> version of LeoCAD.

OK, I'll trust your judgement (and implementation).

>     const float LightDarkControl = Preferences.mAutomateEdgeColor ?
>     Preferences.mPartColorValueLDIndex :
>     LC_SRGB_TO_LINEAR(Preferences.mPartColorValueLDIndex);
>     It's correct. I just think it's confusing (why LC_SRGB_TO_LINEAR for
>     one path but not the other)
> 
> Ok. As we already discussed, this flag serves both HCS and AEC 
> behaviours so based on what is enabled, the appropriate assignment is 
> handled. AEC uses the index directly while HCS uses linear derivative.

Understood; it's just somewhat confusing until examined in depth. It 
pains me to state the following since my early programming jobs were in 
assembler (Z80) on very limited system by today's standards, but I've 
learned that it's usually better to make the code __more__ maintainable 
and easier to follow by others even if it's less efficient. My 
tendencies are still towards the terse and I put real effort into 
controlling that.

**Comment by leozide** [comment]
> Ok, I did not use your patch update code. I opted for the 'Adjusted' flag instead. The color file loads whenever LoadColors() is called, you can see when this is by looking up this call in the source. Once a colour is adjusted, there is no need to adjust it again so the flag allows the lcAdjustStudStyleColors() loop to skip an adjusted colour. However, neither this nor your perceived bug should ever trigger in the current LeoCAD behaviour because lcAdjustStudStyleColors() is only called from lcLoadColorFile() so there should never be an opportunity to adjust an already adjusted colour. I added this flag in the event Leo decides to recognize the !COLOUR LDraw meta command when present in model files in which case I imagine the colour file will not be reloaded. For LPub3D, I use the !COLOUR meta command in my modified version of LeoCAD.

That's interesting, I didn't think anyone used !COLOUR in model files. I don't handle that at all.

**Comment by trevorsandy** [comment]
> That's interesting, I didn't think anyone used !COLOUR in model files. I don't handle that at all.

Yep, this is actually the preferred way to modify a colour vs. manually changing or using an alternative LDConfig.ldr. This command is natively supported by LDView and LDGLite. Perhaps it’s not so interesting for editing/viewing but it is certainly a valid feature for publishing models/instructions where the user would like to tweak part colours. Since before you implemented fade and highlight colours, I modded  my LeoCAD renderer to use it to manage these feature settings from LPub3D. That you have begun to enhance instructions generation in LeoCAD, it could be good to add this capability.

Cheers,

**Comment by leozide** [comment]
> Yep, this is actually the preferred way to modify a colour vs. manually changing or using an alternative LDConfig.ldr. This command is natively supported by LDView and LDGLite. Perhaps it’s not so interesting for editing/viewing but it is certainly a valid feature for publishing models/instructions where the user would like to tweak part colours. Since before you implemented fade and highlight colours, I modded my LeoCAD renderer to use it to manage these feature settings from LPub3D. That you have begun to enhance instructions generation in LeoCAD, it could be good to add this capability.

So when there's a !COLOUR in a model file, does it affect all parts in the model (both before and after the command)? What about submodels?

I think it would be hard for me to change just a few parts.

File: common/lc_colors.cpp [hunk 1]
```diff
@@ -222,22 +222,21 @@ static void lcAdjustStudStyleColors(std::vector<lcColor>& Colors, lcStudStyle St
 	const lcVector4 BlackEdge = lcVector4FromColor(Preferences.mBlackEdgeColor);
 
 	const float ContrastControl = Preferences.mPartEdgeContrast;
-	const float LightDarkControl = Preferences.mAutomateEdgeColor ?
-				Preferences.mPartColorValueLDIndex :
-				LC_SRGB_TO_LINEAR(Preferences.mPartColorValueLDIndex);
+	const float LightDarkControl = LC_SRGB_TO_LINEAR(Preferences.mAutomateEdgeColor ? Preferences.mPartColorValueLDIndex : Preferences.mPartColorValueLDIndex);
```
Comment (rsbx) [inline common/lc_colors.cpp:225]: This particular hunk needs to be reverted; the old code was correct. See my note in the #621 thread for the reason it was done that way.

The other changes appear to be correct (in this first reading).

**Pull request merged**